### PR TITLE
Remove the configuration state we keep in the remote

### PR DESCRIPTION
**Pull request by carlosmn**
There's way to much state which we keep in the remote, which can be changed at run-time and maybe saved awkwardly to the configuration.

As a first step, I've removed `git_remote_{s,g}et_callbacks` and added `git_fetch_options` to keep track of the callbacks, maybe in this PR or later.

My end goal for the remote is to remove `git_remote_save()` and have changes happen on-disc but have the remote's properties be read-only from the point it was looked up and the options for fetch or push be determined via the options you pass. See #2711 for some more on this.

### CODE REVIEW

**Comment by carlosmn** [comment]
I'm not entirely sure about the `set_transport()` change, but as remotes are not long-lived objects, it's probably ok that you pass this in as well.

`git_remote_set_autotag()` is a concession to the way we were behaving with it in `git_remote_save()` but I'm not sure if this is a meaningful thing to do. A user is going to do a `git config` so they won't care, and a GUI or so may not want to rely on our logic for deletion and whatnot.

**Comment by carlosmn** [comment]
The tests should be passing now and we no longer keep any runtime configuration in the remote, so this is ready for review.

**Comment by ethomson** [comment]
I love it when the number of files changed in a PR is roughly equal to the number of files in the repository...

;)

**Comment by carlosmn** [comment]
We do still have `git_remote_set_{fetch,push}_refspecs()` if someone wants to change what the base refspec list is supposed to be. I'm not sure having it is a good idea, but it does add some flexibility where it would be hard without it (though I don't know if you'd actually want to do something like this).

Similarly, I'm not sure I like that `git_remote_create_anonymous()` gets a base refspec. The way these would work with git they do not get any, just the list of refspecs used for the particular fetch or push. I'm not sure if it'd be useful to keep it or better to keep it more git-like and remove the refspec from an anonymous remote.

**Comment by ethomson** [comment]
I really like this.  Big heads up to binding authors, your world is about to get real.  Sorry as per usual.

I trust that since this has been open for a little while and nobody went *nooooooooooo* that this is generally well accepted to not put you in a new world of pain when it comes to bindings.  (Except for the world to adapt to the new API.)  But I'm pinging several of you explicitly just in case you want to complain about this.

/cc @nulltoken 
/cc @jamill 
/cc @arthurschreiber 
/cc @ethomson oh no! :(

**Comment by arthurschreiber** [comment]
Overall, this LGTM. I've not had the time to actually try this out and update the rugged bindings to use this, but I think that it should not be too hard to implement these changes.

**Comment by carlosmn** [comment]
Rebased to take into account the new push reporting.

File: CHANGELOG.md [hunk 1]
```diff
@@ -130,6 +153,39 @@ support for HTTPS connections insead of OpenSSL.
...
+  accordingly:
+
+    * `git_remote_set_url()`, `git_remote_seturl()`
+    * `git_remote_add_fetch()`, `git_remote_add_push()` and
+    * `git_remote_set_autotag()`
+
+* `git_remote_connect()` and `git_remote_prune()` now take a pointer
+  to the callbacks.
+
+* `git_remote_fetch()` and `git_remote_download()` now take a poitner
```
Comment (jamill) [inline CHANGELOG.md]: poitner -> pointer

File: CHANGELOG.md [hunk 2]
```diff
@@ -74,6 +88,15 @@ support for HTTPS connections insead of OpenSSL.
 
 ### API removals
 
+* `git_remote_save()` and `git_remote_clear_refspecs()` has been
```
Comment (jamill) [inline CHANGELOG.md]: nit: has -> have

File: CHANGELOG.md [hunk 3]
```diff
@@ -74,6 +88,15 @@ support for HTTPS connections insead of OpenSSL.
...
+  removed. Remote's configuration is changed via the configuration
+  directly or through a convenience function which performs changes to
+  the configuration directly.
+
+* `git_remote_set_callbacks()`, `git_remote_get_callbacks()` and
+  `git_remote_set_transport()` have been removed a the remote no
```
Comment (jamill) [inline CHANGELOG.md]: a -> and?

**Pull request merged**